File: java/08_029.java
import java.util.HashSet;
import java.util.Set;

/**
 * Checks if a circuit can be formed in a binary matrix
 */
public class MatrixCircuit {
   // Directions: right, down, left, up
   private static final int[][] DIRECTIONS = 
      {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
   
   /**
    * Determines if a circuit can be made from start position
    */
   public static boolean canCompleteCircuit(int[][] matrix, 
         int startRow, int startCol) {
      if (matrix[startRow][startCol] == 1) {
         return false;  // Invalid starting position
      }
      
      int rows = matrix.length;
      int cols = matrix[0].length;
      Set<Pair> visited = new HashSet<>();
      
      return dfs(matrix, startRow, startCol, startRow, startCol, 
            0, visited, rows, cols);
   }
   
   /**
    * DFS to find a valid circuit
    */
   private static boolean dfs(int[][] matrix, int startRow, 
         int startCol, int row, int col, 
         int pathLength, Set<Pair> visited, 
         int rows, int cols) {
      // Complete circuit found (back to start after ≥3 moves)
      if (row == startRow && col == startCol && pathLength >= 3) {
         return true;
      }
      
      visited.add(new Pair(row, col));
      
      for (int[] dir : DIRECTIONS) {
         int newRow = row + dir[0];
         int newCol = col + dir[1];
         
         // Check if move is valid
         if (isValidMove(matrix, newRow, newCol, rows, cols)) {
            Pair newPos = new Pair(newRow, newCol);
            // Allow revisiting start position to complete circuit
            if (!visited.contains(newPos) || 
                (newRow == startRow && newCol == startCol && 
                pathLength >= 3)) {
               
               if (dfs(matrix, startRow, startCol, newRow, newCol,
                     pathLength + 1, visited, rows, cols)) {
                  return true;
               }
            }
         }
      }
      
      // Backtrack
      visited.remove(new Pair(row, col));
      return false;
   }
   
   /**
    * Check if position is valid and not a wall
    */
   private static boolean isValidMove(int[][] matrix, int row, 
         int col, int rows, int cols) {
      return row >= 0 && row < rows && col >= 0 && col < cols && 
            matrix[row][col] == 0;
   }
   
   /**
    * Helper class for coordinate pairs
    */
   static class Pair {
      int row, col;
      
      Pair(int row, int col) {
         this.row = row;
         this.col = col;
      }
      
      @Override
      public boolean equals(Object obj) {
         if (this == obj) return true;
         if (obj == null || getClass() != obj.getClass()) 
            return false;
         Pair pair = (Pair) obj;
         return row == pair.row && col == pair.col;
      }
      
      @Override
      public int hashCode() {
         return 31 * row + col;
      }
   }
   
   /**
    * Test the circuit finder
    */
   public static void main(String[] args) {
      int[][] matrix = {
         {0, 1, 0, 0},
         {0, 0, 0, 1},
         {1, 1, 0, 0},
         {0, 0, 0, 0}
      };
      
      System.out.println("Circuit possible: " + 
            canCompleteCircuit(matrix, 0, 0));  // true
   }
}

File: python/08_029.py
def can_complete_circuit(matrix, start_row, start_col):
  if matrix[start_row][start_col] == 1:
     return False  # Invalid starting position
  
  rows, cols = len(matrix), len(matrix[0])
  directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]  # r,d,l,u
  visited = set()
  
  def dfs(row, col, path_length):
     # If we've returned to start after visiting 3+ cells
     if ((row, col) == (start_row, start_col) and 
         path_length >= 3):
        return True
     
     visited.add((row, col))
     
     for dr, dc in directions:
        new_row, new_col = row + dr, col + dc
        
        # Check if new position is valid and not visited
        if ((0 <= new_row < rows and 
             0 <= new_col < cols and 
             matrix[new_row][new_col] == 0 and 
             (new_row, new_col) not in visited) or 
            ((new_row, new_col) == (start_row, start_col) and 
             path_length >= 3)):
           
           if dfs(new_row, new_col, path_length + 1):
              return True
     
     # Backtrack
     visited.remove((row, col))
     return False
  
  return dfs(start_row, start_col, 0)

# Test case
matrix = [
  [0, 1, 0, 0],
  [0, 0, 0, 1],
  [1, 1, 0, 0],
  [0, 0, 0, 0]
]
print(can_complete_circuit(matrix, 0, 0))  # Should return True